### ISSUE TITLE
Question/Concern - Conversion to Backdrop Breaks Entity API Functionality

### ISSUE THREAD

**Issue by rbargerhuff** [backdrop-contrib/coder_upgrade#75]
Hello there,

When converting in-house modules or contrib modules to Backdrop, we have found that references to several classes included in Drupal 7's EntityAPI module will ultimately break the converted module. For example:

EntityApiController in Drupal 7's Entity API module has been changed to EntityPlusController in Backdrop's Entity Plus module.

We are aware that there are provided hooks by Coder Upgrade that allows targeted code to be renamed. However, we suggest, since the Entity API is such an integral part of Drupal 7, that Coder Upgrade should be upgraded to handle the above conversion.

Without this conversion, functionality provided by 3rd party modules will ultimately break.

Thoughts?

**Comment by argiepiano** [comment]
Hi @rbargerhuff . I'm one of the maintainers of Entity Plus. I agree that these class conversions should be included in Coder Update. I'm wondering if providing a wrapper class named EntityApiController extending EntityPlusController within the Entity plus module may be helpful too? That said, a lot of the API functions in entity plus have been renamed with _plus_  to avoid collisions with core, so there may be some additional work needed to convert those.

**Comment by rbargerhuff** [comment]
Hello there @argiepiano . Thank you so much for your reply. So we went ahead and modified our fork of coder_upgrade to make the adjustments needed for coder_upgrade to take care of the conversion. We were waiting on a response from the maintainer(s) of coder_upgrade before submitting our resolution.

As far as what you proposed, I feel that could be a partial solution. Whether coder_upgrade is updated depends on EntityPlus, not being part of core, is where the proverbial line is drawn.

This is the code we added to conversions/regex.inc beginning @ line 99.

```
//<editor-fold desc="handle the renaming of classes associated with the EntityAPI module in Drupal 7 to the core module EntityPlus">
  $entityAPIFrom = [
    'EntityAPITestCase',
    'EntityAPIController',
    'EntityAPIControllerExportable',
    'EntityAPIControllerInterface',
    'EntityAPIControllerRevisionableInterface',
    'entity_views_handler_area_entity',
    'entity_views_handler_field_boolean',
    'entity_views_handler_field_date',
    'entity_views_handler_field_duration',
    'entity_views_handler_field_entity',
    'entity_views_handler_field_field',
    'entity_views_handler_field_numeric',
    'entity_views_handler_field_options',
    'entity_views_handler_field_text',
    'entity_views_handler_field_uri',
    'entity_views_handler_relationship_by_bundle',
    'entity_views_handler_relationship',
    'entity_views_plugin_row_entity_view'
  ];

  $entityAPITo = [
    'EntityPlusMetadataTestCase',
    'EntityPlusController',
    'EntityPlusControllerExportable',
    'EntityPlusControllerInterface',
    'EntityPlusControllerRevisionableInterface',
    'entity_plus_views_handler_area_entity',
    'entity_plus_views_handler_field_boolean',
    'entity_plus_views_handler_field_date',
    'entity_plus_views_handler_field_duration',
    'entity_plus_views_handler_field_entity',
    'entity_plus_views_handler_field_field',
    'entity_plus_views_handler_field_numeric',
    'entity_plus_views_handler_field_options',
    'entity_plus_views_handler_field_text',
    'entity_plus_views_handler_field_uri',
    'entity_plus_views_handler_relationship_by_bundle',
    'entity_plus_views_handler_relationship',
    'entity_plus_views_plugin_row_entity_view'
  ];

  // add the from collection associated with the EntityAPI.
  foreach ($entityAPIFrom as $ef) {
    $from[] = "/" . $ef . "/";
  }

  // add the to collection associated with the EntityAPI.
  foreach ($entityAPITo as $et) {
    $to[] = $et;
  }
  //</editor-fold>
```

**Comment by bugfolder** [comment]
This looks like a nice addition. If any of these conversions are made, then Entity Plus would need to be added as a dependency to the .info file, and probably a warning message posted to alert the user of the new dependency.

@docwilmot, your thoughts?

**Comment by rbargerhuff** [comment]
@bugfolder Yup! We were thinking about that very hiccup. We would need to detect that the EntityAPI module was being used and then add a dependency like you said and then a message after the conversion indicating that EntityPlus is now required.

**Comment by argiepiano** [comment]
I'm not familiar with Coder Upgrade, but I see in the API definitions that it's possible for other modules to contribute their own conversions. Would it make sense to provide an upgrade implementation as a separate submodule of Entity Plus, rather than adding the code here? And if so, can Coder Upgrade warn someone that they would need to enable such submodule if it detects the use of Entity API?

**Comment by rbargerhuff** [comment]
We were thinking the same thing before we modified Coder Upgrade that using the provided hooks could accomplish the renaming. Our argument was that EntityAPI/EntityPlus should be an exception since EntityAPI is such an integral part of Drupal 7 (Close to 400,000 sites are using EntityAPI).

However, I can see the argument against why Coder Upgrade should not be handling the specific changes in question.

To answer your question tho, if EntityPlus were to include a submodule that handles the conversion that would be ideal. Being new to this process, am not sure how that would work and could only provide little input.

**Comment by argiepiano** [comment]
> Our argument was that EntityAPI/EntityPlus should be an exception since EntityAPI is such an integral part of Drupal 7 (Close to 400,000 sites are using EntityAPI).

Yeah, with those numbers I can see why it would make sense to make this as straightforward as possible by including this conversion by default.

**Comment by docwilmot** [comment]
Very delayed response, sorry. This would be very helpful, but by principle, it should be implemented in Entity Plus I'd say. But I see the argument.
And (related) maybe we could implement code for the new [Plugin Manager](https://github.com/backdrop-contrib/plugin_manager) module as well to convert CTools plugins?
